### PR TITLE
Caching MNIST dataset for testing

### LINKED ISSUE #859: test fails with `_lazy_train_dataloader`
## 🐛 Bug

The tests are time-to-time randomly failing across all platforms (but mainly macOS and Win) with following message:

```
_____________________________ test_lbfgs_cpu_model _____________________________

self = LightningTestModel(
  (c_d1): Linear(in_features=784, out_features=1000, bias=True)
  (c_d1_bn): BatchNorm1d(1000, eps...ats=True)
  (c_d1_drop): Dropout(p=0.2, inplace=False)
  (c_d2): Linear(in_features=1000, out_features=10, bias=True)
)

    def _get_data_loader(self):
        try:
>           value = getattr(self, attr_name)

pytorch_lightning/core/decorators.py:16: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = LightningTestModel(
  (c_d1): Linear(in_features=784, out_features=1000, bias=True)
  (c_d1_bn): BatchNorm1d(1000, eps...ats=True)
  (c_d1_drop): Dropout(p=0.2, inplace=False)
  (c_d2): Linear(in_features=1000, out_features=10, bias=True)
)
name = '_lazy_train_dataloader'

    def __getattr__(self, name):
        if '_parameters' in self.__dict__:
            _parameters = self.__dict__['_parameters']
            if name in _parameters:
                return _parameters[name]
        if '_buffers' in self.__dict__:
            _buffers = self.__dict__['_buffers']
            if name in _buffers:
                return _buffers[name]
        if '_modules' in self.__dict__:
            modules = self.__dict__['_modules']
            if name in modules:
                return modules[name]
        raise AttributeError("'{}' object has no attribute '{}'".format(
>           type(self).__name__, name))
E       AttributeError: 'LightningTestModel' object has no attribute '_lazy_train_dataloader'
```

### To Reproduce

https://github.com/Borda/pytorch-lightning/runs/448238658

### PR DESCRIPTION
**Pull request by djbyrne**
* Added MNIST datset to the tests directory

* Caches dataset based off hash of the test.pt file

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CHANGELOG.md)?

## What does this PR do?
Fixes #859 .
Caches the MNIST dataset. This will prevent the CI/CD testing from crashing when the link to the original dataset is unavailable and will speed up CI/CD as the pipeline will not have to re-download files.

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
is it not possible to download it once and reuse? otherwise the package will grow by 100+ MB for only the tests and most users don't want that.

**Comment by williamFalcon** [comment]
thanks for the PR! we can’t add a dataset to the framework. @Borda

**Comment by Borda** [comment]
We are not adding them to package, it is just in CI test time, e.g. running PRs... Let review the PR =)

**Comment by williamFalcon** [comment]
i see binaries in the /test folder

**Comment by williamFalcon** [comment]
https://github.com/djbyrne/pytorch-lightning/tree/enhancement/lazy_dataloader/tests/models/mnist/MNIST

**Comment by Borda** [comment]
> i see binaries in the /test folder

Yeah the was probably misunderstanding when we talked about caching lol

**Comment by williamFalcon** [comment]
https://pl-public-data.s3.amazonaws.com/MNIST/processed/test.pt
https://pl-public-data.s3.amazonaws.com/MNIST/processed/training.pt

https://pl-public-data.s3.amazonaws.com/MNIST/raw/t10k-images-idx3-ubyte
https://pl-public-data.s3.amazonaws.com/MNIST/raw/t10k-images-idx3-ubyte.gz
https://pl-public-data.s3.amazonaws.com/MNIST/raw/t10k-labels-idx1-ubyte
https://pl-public-data.s3.amazonaws.com/MNIST/raw/t10k-labels-idx1-ubyte.gz
https://pl-public-data.s3.amazonaws.com/MNIST/raw/train-images-idx3-ubyte
https://pl-public-data.s3.amazonaws.com/MNIST/raw/train-images-idx3-ubyte.gz
https://pl-public-data.s3.amazonaws.com/MNIST/raw/train-labels-idx1-ubyte
https://pl-public-data.s3.amazonaws.com/MNIST/raw/train-labels-idx1-ubyte.gz

**Comment by djbyrne** [comment]
> > i see binaries in the /test folder
> 
> Yeah the was probably misunderstanding when we talked about caching lol

@Borda  apologies, I misunderstood the request. Ill remove the MNIST files from the directory.

**Comment by djbyrne** [comment]
It looks like the caching of the pip files is only being done for Linux. Should I add in caching of the locations for Windows and MacOS as well?

**Comment by williamFalcon** [comment]
@Borda good to go?

**Comment by williamFalcon** [comment]
@djbyrne can you subclass MNIST dataset for tests only and add the urls we have?

**Comment by djbyrne** [comment]
@williamFalcon this is just to override the resources list of links to the dataset right. This is different to the request to have an MNIST class decoupled from torchvision that @Borda mentioned?

**Comment by Borda** [comment]
Pls do not remove the blank lines, then it shall be fine for Linux tests
It would be nice to have the caching also for other platforms...

File: .github/workflows/ci-testing.yml [hunk 1]
```diff
@@ -45,15 +52,13 @@ jobs:
...
-
     - name: Lint and Tests
       # env:
       #   TOXENV: py${{ matrix.python-version }}
       run: |
         # tox --sitepackages
         flake8 .
         coverage run --source pytorch_lightning -m py.test pytorch_lightning tests pl_examples -v --doctest-modules --junitxml=junit/test-results-${{ runner.os }}-${{ matrix.python-version }}.xml
-
     - name: Upload pytest test results
```
Comment (Borda) [inline .github/workflows/ci-testing.yml]:
```suggestion

    - name: Upload pytest test results
```

File: .github/workflows/ci-testing.yml [hunk 2]
```diff
@@ -68,7 +73,6 @@ jobs:
         python setup.py check --metadata --strict
         python setup.py sdist
         twine check dist/*
-
     #- name: Try install package
```
Comment (Borda) [inline .github/workflows/ci-testing.yml]:
```suggestion

    #- name: Try install package
```

File: .github/workflows/ci-testing.yml [hunk 3]
```diff
@@ -45,7 +52,6 @@ jobs:
         python --version
         pip --version
         pip list
-
     - name: Lint and Tests
```
Comment (Borda) [inline .github/workflows/ci-testing.yml]:
```suggestion

    - name: Lint and Tests
```

**Pull request merged**